p2p/nodestate/nodestate.go

```diff
@@ -646,55 +662,100 @@ func (ns *NodeStateMachine) SetState(n *enode.Node, setFlags, resetFlags Flags,
-		if changed&sub.mask != 0 {
-			sub.callback(n, Flags{mask: oldState & sub.mask, setup: ns.setup}, Flags{mask: newState & sub.mask, setup: ns.setup})
+	callback := func() {
+		for _, sub := range ns.stateSubs {
+			if changed&sub.mask != 0 {
+				sub.callback(n, Flags{mask: oldState & sub.mask, setup: ns.setup}, Flags{mask: newState & sub.mask, setup: ns.setup}, caller)
+			}
 		}
-	}
-	if newState == 0 {
-		// call field subscriptions for discarded fields
-		for i, v := range node.fields {
-			if v != nil {
-				f := ns.fields[i]
-				if len(f.subs) > 0 {
-					for _, cb := range f.subs {
-						cb(n, Flags{setup: ns.setup}, v, nil)
+		if newState == 0 {
+			// call field subscriptions for discarded fields
+			for i, v := range node.fields {
+				if v != nil {
+					if debugPrints {
+						fmt.Println("discardField", n.ID(), ns.setup.fields[i].name, v)
+					}
+					f := ns.fields[i]
+					if len(f.subs) > 0 {
+						for _, cb := range f.subs {
+							cb(n, Flags{setup: ns.setup}, v, nil, caller)
+						}
 					}
 				}
 			}
 		}
 	}
+	callNow := caller == nil
+	if callNow {
+		caller = &Caller{}
+	} else {
+		caller.pending = append(caller.pending, callback)
+		ns.callbackCount++
+	}
+	ns.lock.Unlock()
+	// call state update subscription callbacks without holding the mutex
+	if callNow {
+		callback()
+		ns.processCallbacks(caller)
+	}
+}
+
+// processCallbacks runs pending callbacks of a given node in a guaranteed correct order.
+// Callbacks resulting from a state/field change performed in a previous callback are always
+// put at the end of the pending list and therefore processed after all callbacks resulting
+// from the previous state/field change.
+func (ns *NodeStateMachine) processCallbacks(caller *Caller) {
+	for len(caller.pending) != 0 {
+		list := caller.pending
+		for _, cb := range list {
+			cb()
+		}
+		ns.lock.Lock()
+		caller.pending = caller.pending[len(list):]
+		ns.callbackCount -= len(list)
+		if ns.callbackCount == 0 {
+			ns.callbackWait.Signal()
+		}
+		ns.lock.Unlock()
+	}
 }
 
 // offlineCallbacks calls state update callbacks at startup or shutdown
 func (ns *NodeStateMachine) offlineCallbacks(start bool) {
+	ns.lock.Lock()
+	caller := &Caller{}
 	for _, cb := range ns.offlineCallbackList {
-		for _, sub := range ns.stateSubs {
-			offState := offlineState & sub.mask
-			onState := cb.state & sub.mask
-			if offState != onState {
-				if start {
-					sub.callback(cb.node, Flags{mask: offState, setup: ns.setup}, Flags{mask: onState, setup: ns.setup})
-				} else {
-					sub.callback(cb.node, Flags{mask: onState, setup: ns.setup}, Flags{mask: offState, setup: ns.setup})
-				}
-			}
-		}
-		for i, f := range cb.fields {
-			if f != nil && ns.fields[i].subs != nil {
-				for _, fsub := range ns.fields[i].subs {
+		cb := cb
+		callback := func() {
+			for _, sub := range ns.stateSubs {
+				offState := offlineState & sub.mask
+				onState := cb.state & sub.mask
+				if offState != onState {
```

Comment:
Reverse the check, `continue`, un-indent, IMO

Reply:
done